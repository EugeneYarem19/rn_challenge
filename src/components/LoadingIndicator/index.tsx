import React from "react";
import { ActivityIndicator, ActivityIndicatorProps, } from "react-native";

import { palette, } from "@theme";

export const LoadingIndicator: React.FC<ActivityIndicatorProps> = ({ color = palette.accentColor, size = "large", }) => (
  <ActivityIndicator size={size} color={color} />
);
