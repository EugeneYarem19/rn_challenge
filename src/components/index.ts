export * from "./LoadingIndicator";
export * from "./SearchBar";
