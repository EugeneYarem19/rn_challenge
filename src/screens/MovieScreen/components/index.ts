export * from "./DetailsItem";
export * from "./MovieScreenComponent";
export * from "./Rating";
