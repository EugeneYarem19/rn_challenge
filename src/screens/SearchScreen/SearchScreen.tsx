import React from "react";
import { FlatList, Text, View, } from "react-native";

import { Movie, } from "@redux";

import { styles, } from "./styles";

interface SearchCallback {
  (): void;
}

interface Props {
  fetchMore: SearchCallback;
  foundMovies: Movie[];
  isFetchingMore: boolean;
  isSearching: boolean;
  searchErrorMessage: string;
}

export const SearchScreen: React.FC<Props> = ({ fetchMore, foundMovies, isFetchingMore, isSearching, searchErrorMessage, }): JSX.Element => (
  <View style={styles.screen}>
    {searchErrorMessage !== "" ? (
      <Text>{searchErrorMessage}</Text>
    ) : (
      <FlatList
        data={foundMovies}
        keyExtractor={(item) => item.id}
        renderItem={({ item, }) => (
          <View>
            <Text>{item.title}</Text>
          </View>
        )}
      />
    )}
  </View>
);
