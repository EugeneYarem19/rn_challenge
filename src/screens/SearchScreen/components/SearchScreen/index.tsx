import React from "react";
import { FlatList, Text, View, } from "react-native";

import { LoadingIndicator, } from "@components";

import { ISearchScreen, } from "./types";
import { styles, } from "./styles";

export const SearchScreen: React.FC<ISearchScreen> = ({ fetchMore, foundMovies, isFetchingMore, isSearching, renderItem, searchErrorMessage, }): JSX.Element => (
  <View style={styles.screen}>
    {isSearching ? (
      <LoadingIndicator />
    ) : searchErrorMessage !== "" ? (
      <View style={styles.errorMessageBlock}>
        <Text style={styles.errorMessageText}>{searchErrorMessage}</Text>
      </View>
    ) : (
      <FlatList
        data={foundMovies}
        keyExtractor={(item) => item.id}
        renderItem={renderItem}
        style={styles.list}
        contentContainerStyle={styles.listContentContainer}
        onEndReached={fetchMore}
        onEndReachedThreshold={0.3}
        ListFooterComponent={isFetchingMore ? <LoadingIndicator /> : null}
      />
    )}
  </View>
);
