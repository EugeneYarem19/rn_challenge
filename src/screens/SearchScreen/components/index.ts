export * from "./SearchResultItem";
export * from "./SearchScreen";
