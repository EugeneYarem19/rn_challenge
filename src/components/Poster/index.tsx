import React from "react";
import ProgressImage from "react-native-image-progress";
import { Icon, } from "react-native-material-ui";
import { View, } from "react-native";

import { LoadingIndicator, } from "../LoadingIndicator";

import { IPoster, ImageSize, } from "./types";
import { styles, } from "./styles";
import { usePoster, } from "./hooks";

export const Poster: React.FC<IPoster> = ({
  poster,
  size = ImageSize.small,
  smallOnEmptyPoster = true,
  smallOnError = true,
  style,
}) => {
  const sizeStyle = size === ImageSize.small ? styles.small : styles.large;
  const { loadError, onError, } = usePoster();

  const renderPlaceholder = (additionalSizedStyle: typeof style) => (
    <View style={[styles.placeholder, additionalSizedStyle, style,]}>
      <Icon name="wallpaper" size={50} />
    </View>
  );

  const renderPosterIcon = () => {
    if (loadError) {
      return renderPlaceholder(smallOnError ? styles.small : sizeStyle);
    } else if (!poster) {
      return renderPlaceholder(smallOnEmptyPoster ? styles.small : sizeStyle);
    } else {
      return (
        <ProgressImage
          source={{ uri: poster, }}
          indicator={LoadingIndicator}
          onError={onError}
          resizeMode="contain"
          style={[styles.poster, sizeStyle, style,]}
          threshold={0}
        />
      );
    }
  };

  return renderPosterIcon();
};
