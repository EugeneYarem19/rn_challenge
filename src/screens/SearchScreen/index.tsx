import React, { useCallback, } from "react";
import { useDispatch, useSelector, } from "react-redux";

import { Movie, State, actions, } from "@redux";

import { SearchScreen as SearchScreenComponent, } from "./SearchScreen";
import { SearchResultItem, } from "./components";

export const SearchScreen = (): JSX.Element => {
  const dispatch = useDispatch();
  const foundMovies = useSelector((state: State) => state.foundMovies);
  const isFetchingMore = useSelector((state: State) => state.isFetchingMore);
  const isSearching = useSelector((state: State) => state.isSearching);
  const isThatsAll = useSelector((state: State) => state.isThatsAll);
  const searchErrorMessage = useSelector((state: State) => state.searchErrorMessage);

  const fetchMore = useCallback(() => {
    if (!isFetchingMore && !isThatsAll) {
      console.warn("fetch");
      dispatch(actions.fetchMore());
    }
  }, [dispatch, isFetchingMore, isThatsAll,]);
  const renderItem = useCallback(
    ({ item: { id, title, poster, }, }: { item: Movie }): JSX.Element => (
      <SearchResultItem title={title} testID={id} poster={poster} onPress={() => dispatch(actions.fetchMovie(id))} />
    ),
    [dispatch,]
  );

  return (
    <SearchScreenComponent
      fetchMore={fetchMore}
      foundMovies={foundMovies}
      isFetchingMore={isFetchingMore}
      isSearching={isSearching}
      renderItem={renderItem}
      searchErrorMessage={searchErrorMessage}
    />
  );
};
