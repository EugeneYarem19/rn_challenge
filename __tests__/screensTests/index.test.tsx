import React from "react";
import configureStore from "redux-mock-store";
import { Provider, } from "react-redux";
import { render, fireEvent, } from "@testing-library/react-native";

import { SearchScreen, } from "@screens";
import { SearchResultItem, } from "@src/screens/SearchScreen/components";

import snapshotTest from "../snapshotTestUtil";
import { mockedData, } from "../mockedData";

const mockStore = configureStore([]);

describe("screens tests", () => {
  test("SearchResultItem must renders correctly with poster", () => {
    snapshotTest(<SearchResultItem testID={mockedData.id} title={mockedData.title} poster={mockedData.movies[0].poster} onPress={() => null} />);
  });

  test("SearchResultItem must renders correctly without poster", () => {
    snapshotTest(<SearchResultItem testID={mockedData.id} title={mockedData.title} poster="" onPress={() => null} />);
  });

  test("SearchResultItem must be clickable", () => {
    const onPress = jest.fn();
    const { getByTestId, } = render(<SearchResultItem testID={mockedData.id} title={mockedData.title} poster="" onPress={onPress} />);

    fireEvent.press(getByTestId(mockedData.id));

    expect(onPress).toHaveBeenCalled();
  });

  test("SearchScreen must render loader if searching even if searchErrorMessage and movies presented", () => {
    const store = mockStore({
      isFetchingMore: false,
      isSearching: true,
      isThatsAll: false,
      searchErrorMessage: mockedData.errorMessage,
      foundMovies: mockedData.movies,
    });
    snapshotTest(
      <Provider store={store}>
        <SearchScreen />
      </Provider>
    );
  });

  test("SearchScreen must display searchErrorMessage if error even if movies presented", () => {
    const store = mockStore({
      isFetchingMore: false,
      isSearching: false,
      isThatsAll: false,
      searchErrorMessage: mockedData.errorMessage,
      foundMovies: mockedData.movies,
    });
    snapshotTest(
      <Provider store={store}>
        <SearchScreen />
      </Provider>
    );
  });

  test("SearchScreen must render movies list", () => {
    const store = mockStore({
      isFetchingMore: false,
      isSearching: false,
      isThatsAll: false,
      searchErrorMessage: "",
      foundMovies: mockedData.movies,
    });
    snapshotTest(
      <Provider store={store}>
        <SearchScreen />
      </Provider>
    );
  });

  test("SearchScreen must render movies list and bottom loader", () => {
    const store = mockStore({
      isFetchingMore: true,
      isSearching: false,
      isThatsAll: false,
      searchErrorMessage: "",
      foundMovies: mockedData.movies,
    });
    snapshotTest(
      <Provider store={store}>
        <SearchScreen />
      </Provider>
    );
  });
});
