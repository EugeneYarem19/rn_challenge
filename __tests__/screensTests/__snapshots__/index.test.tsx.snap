// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`screens tests SearchResultItem must renders correctly with poster 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flex": 1,
      "flexDirection": "column-reverse",
      "height": 200,
      "marginBottom": 20,
      "opacity": 1,
      "overflow": "hidden",
    }
  }
  testID="t1"
>
  <View
    style={
      Object {
        "backgroundColor": "#1e3166",
        "borderRadius": 5,
        "flexDirection": "row",
        "height": 165,
        "paddingHorizontal": 16,
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "http://",
        }
      }
      style={
        Object {
          "borderRadius": 8,
          "height": 180,
          "marginTop": -35,
          "width": 122,
        }
      }
    />
    <Text
      style={
        Object {
          "color": "#ccc",
          "flex": 1,
          "fontSize": 20,
          "fontWeight": "bold",
          "marginHorizontal": 16,
          "marginVertical": 16,
        }
      }
    >
      Test
    </Text>
  </View>
</View>
`;

exports[`screens tests SearchResultItem must renders correctly without poster 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flex": 1,
      "flexDirection": "column-reverse",
      "height": 200,
      "marginBottom": 20,
      "opacity": 1,
      "overflow": "hidden",
    }
  }
  testID="t1"
>
  <View
    style={
      Object {
        "backgroundColor": "#1e3166",
        "borderRadius": 5,
        "flexDirection": "row",
        "height": 165,
        "paddingHorizontal": 16,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#131d3d",
          "borderRadius": 8,
          "flexDirection": "column",
          "height": 180,
          "justifyContent": "center",
          "marginTop": -35,
          "width": 122,
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "rgba(0, 0, 0, 0.54)",
              "fontSize": 50,
            },
            null,
            Object {
              "fontFamily": "Material Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        î†¼
      </Text>
    </View>
    <Text
      style={
        Object {
          "color": "#ccc",
          "flex": 1,
          "fontSize": 20,
          "fontWeight": "bold",
          "marginHorizontal": 16,
          "marginVertical": 16,
        }
      }
    >
      Test
    </Text>
  </View>
</View>
`;

exports[`screens tests SearchScreen must display searchErrorMessage if error even if movies presented 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#131d3d",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#1e3166",
        "borderRadius": 5,
        "flexDirection": "row",
        "marginHorizontal": 16,
        "marginTop": 25,
        "minHeight": 60,
        "padding": 16,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#e59f2b",
          "fontSize": 21,
          "fontWeight": "bold",
        }
      }
    >
      Error
    </Text>
  </View>
</View>
`;

exports[`screens tests SearchScreen must render loader if searching even if searchErrorMessage and movies presented 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#131d3d",
      "flex": 1,
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#e59f2b"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;

exports[`screens tests SearchScreen must render movies list 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#131d3d",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ListFooterComponent={null}
    contentContainerStyle={
      Object {
        "paddingBottom": 25,
      }
    }
    data={
      Array [
        Object {
          "id": "t1",
          "poster": "http://",
          "title": "Test",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.3}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "paddingHorizontal": 16,
        "paddingTop": 25,
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "flex": 1,
              "flexDirection": "column-reverse",
              "height": 200,
              "marginBottom": 20,
              "opacity": 1,
              "overflow": "hidden",
            }
          }
          testID="t1"
        >
          <View
            style={
              Object {
                "backgroundColor": "#1e3166",
                "borderRadius": 5,
                "flexDirection": "row",
                "height": 165,
                "paddingHorizontal": 16,
              }
            }
          >
            <Image
              source={
                Object {
                  "uri": "http://",
                }
              }
              style={
                Object {
                  "borderRadius": 8,
                  "height": 180,
                  "marginTop": -35,
                  "width": 122,
                }
              }
            />
            <Text
              style={
                Object {
                  "color": "#ccc",
                  "flex": 1,
                  "fontSize": 20,
                  "fontWeight": "bold",
                  "marginHorizontal": 16,
                  "marginVertical": 16,
                }
              }
            >
              Test
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`screens tests SearchScreen must render movies list and bottom loader 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#131d3d",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ListFooterComponent={<LoadingIndicator />}
    contentContainerStyle={
      Object {
        "paddingBottom": 25,
      }
    }
    data={
      Array [
        Object {
          "id": "t1",
          "poster": "http://",
          "title": "Test",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.3}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "paddingHorizontal": 16,
        "paddingTop": 25,
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "flex": 1,
              "flexDirection": "column-reverse",
              "height": 200,
              "marginBottom": 20,
              "opacity": 1,
              "overflow": "hidden",
            }
          }
          testID="t1"
        >
          <View
            style={
              Object {
                "backgroundColor": "#1e3166",
                "borderRadius": 5,
                "flexDirection": "row",
                "height": 165,
                "paddingHorizontal": 16,
              }
            }
          >
            <Image
              source={
                Object {
                  "uri": "http://",
                }
              }
              style={
                Object {
                  "borderRadius": 8,
                  "height": 180,
                  "marginTop": -35,
                  "width": 122,
                }
              }
            />
            <Text
              style={
                Object {
                  "color": "#ccc",
                  "flex": 1,
                  "fontSize": 20,
                  "fontWeight": "bold",
                  "marginHorizontal": 16,
                  "marginVertical": 16,
                }
              }
            >
              Test
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <ActivityIndicator
          animating={true}
          color="#e59f2b"
          hidesWhenStopped={true}
          size="large"
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
